File: C5/C5.cpp
#include <iostream>
#include <cmath>
#include <queue>

using namespace std;

struct Ballon
{
    float r;
    int pos;
    int max_r;
};

Ballon ballons[100000];
int main()
{
    int n, pos, max_r, comp_idx = 0;
    double temp_r, init_r;
    double res_r;
    
    cin >> n;
    for (int i = 0; i < n; i++)
    {
        cin >> pos >> max_r;
        ballons[i].pos = pos;
        ballons[i].max_r = max_r;
    }

    cout << fixed;
    cout.precision(3);

    ballons[0].r = ballons[0].max_r;
    res_r = (double)ballons[0].max_r;
    cout << res_r << '\n';

    for (int i = 1; i < n; i++)
    {   
        init_r = ballons[i].max_r;
        
        if(ballons[comp_idx].pos + ballons[comp_idx].r > ballons[i].pos - ballons[i].max_r)
        {
            temp_r = pow((ballons[i].pos - ballons[comp_idx].pos), 2) / (4*ballons[comp_idx].r);
            init_r = min(temp_r, init_r);
            double comp_right = ballons[comp_idx].pos + ballons[comp_idx].r;
            double i_right = ballons[i].pos + init_r;
            if(comp_right < i_right)
                comp_idx = i;
            if(comp_idx != i-1)
            {
                temp_r = pow((ballons[i].pos - ballons[i-1].pos), 2) / (4*ballons[i-1].r);
                init_r = min(temp_r, init_r);
            }
            
        }
        cout << floor(init_r*1000)/1000 << '\n';
        ballons[i].r = floor(init_r*1000)/1000;
    }
}

File: C5/C5.m
%% whole test
test_num = 1;
while 1
    disp(num2str(test_num));
    %% test case generation
    n = 10;
    d = zeros(10, 3);
    d2 = zeros(10, 3);
    for idx = 1:10
        d(idx, 1) = randi(100); % pos
        d(idx, 2) = 0; % r
        d(idx, 3) = randi(100); %max_r
    end
    d = sortrows(d, 1);
    d2 = d;
    %% brute force
    %first ballon is always max
    d(1, 2) = d(1, 3);
    d2(1, 2) = d(1, 3);
    br_max_r = 0;
    op_max_r = 0;
    temp = 0;
    temp_r = 0;
    comp_idx = 1;
    flag = 0;
    for idx = 2:10
        br_max_r = d(idx, 3);
        op_max_r = d(idx, 3);
        %brute force calculation
        for i = idx-1 : -1 : 1
            %if two circle can meet
            if d(i,1)+d(i,2) > d(idx,1) - d(idx,3)
                %brute force calculation
                temp = ((d(idx,1)-d(i,1)).^2) ./ (4*d(i,2));
                temp = fix(temp*10^3)./10^3;
                br_max_r = min(br_max_r, temp);  
            end
        end
        %brute force
        d(idx, 2) = br_max_r;
        %optimize calculation
        if d(comp_idx,1)+d(comp_idx,2) > d(idx,1) - d(idx,3)
            temp_r = ((d(idx,1)-d(comp_idx,1)).^2) ./ (4*d(comp_idx,2));
            temp_r = fix(temp_r*10^3) ./ 10^3;
            op_max_r = min(temp_r, op_max_r);
            if (d(comp_idx, 1) + d(comp_idx, 2)) < (d(idx,1) + op_max_r)
                comp_idx = idx;
            end
            if comp_idx ~= idx-1
                temp_r = ((d(idx,1)-d(idx-1,1)).^2) ./ (4*d(idx-1,2));
                temp_r = fix(temp_r*10^3) ./ 10^3;
                op_max_r = min(temp_r, op_max_r);
            end
        end
        d2(idx, 2) = op_max_r;
        if d(idx,2) ~= d2(idx,2)
            flag = 1;
            break
        end
    end
        
    if flag ==1
        break
    end
%%
    test_num = test_num + 1;
end

%%
% disp([num2str(n) ' ' num2str(n)]);
% for idx = 1:n
%     disp([num2str(d(idx, 2)) ' ' num2str(d(idx, 1))]);
% end